File: app/src/main/java/com/reteno/sample/fragments/database/BaseReadAdapter.java
package com.reteno.sample.fragments.database;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import java.util.List;

import kotlin.Pair;

abstract class BaseReadAdapter<Model, ViewHolderBinding extends ViewBinding, ViewHolder extends BaseReadViewHolder<Model, ViewHolderBinding>> extends RecyclerView.Adapter<ViewHolder> {

    protected List<Model> items;
    protected final ViewHolderListener onExpandCollapseClickListener;

    BaseReadAdapter(ViewHolderListener listener) {
        this.onExpandCollapseClickListener = listener;
    }

    void setItems(List<Model> newItems) {
        this.items = newItems;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public abstract ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType);

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Model item = items.get(position);
        initListeners(holder.bindingHolder);
        holder.bind(item);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    abstract void initListeners(ViewHolderBinding binding);
}

File: app/src/main/java/com/reteno/sample/fragments/database/BaseReadAdapter.kt
package com.reteno.sample.fragments.database

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding

internal abstract class BaseReadAdapter<Model : Any, ViewHolderBinding : ViewBinding, ViewHolder : BaseReadViewHolder<Model, ViewHolderBinding>>(
    @JvmField protected val onExpandCollapseClickListener: ViewHolderListener
) : RecyclerView.Adapter<ViewHolder>() {

    private var itemsList: List<Model>? = null

    fun setItems(newItems: List<Model>?) {
        itemsList = newItems
        notifyDataSetChanged()
    }

    abstract override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = itemsList!![position]
        initListeners(holder.bindingHolder)
        holder.bind(item)
    }

    override fun getItemCount(): Int {
        return itemsList!!.size
    }

    abstract fun initListeners(binding: ViewHolderBinding)
}
